diff --git a/lottie/details/lottie_frame_provider_cached.cpp b/lottie/details/lottie_frame_provider_cached.cpp
index 3a47bc0..3689df5 100644
--- a/lottie/details/lottie_frame_provider_cached.cpp
+++ b/lottie/details/lottie_frame_provider_cached.cpp
@@ -26,9 +26,9 @@ FrameProviderCached::FrameProviderCached(
 		}
 	} else {
 		_direct.setInformation({
-			.size = _cache.originalSize(),
-			.frameRate = _cache.frameRate(),
-			.framesCount = _cache.framesCount(),
+			_cache.originalSize(),
+			_cache.frameRate(),
+			_cache.framesCount(),
 		});
 	}
 }
diff --git a/lottie/details/lottie_frame_provider_cached_multi.cpp b/lottie/details/lottie_frame_provider_cached_multi.cpp
index 0bf5cf1..e075783 100644
--- a/lottie/details/lottie_frame_provider_cached_multi.cpp
+++ b/lottie/details/lottie_frame_provider_cached_multi.cpp
@@ -73,9 +73,9 @@ FrameProviderCachedMulti::FrameProviderCachedMulti(
 	}
 	if (!_direct.loaded()) {
 		_direct.setInformation({
-			.size = _caches.front().originalSize(),
-			.frameRate = _caches.front().frameRate(),
-			.framesCount = ranges::accumulate(
+			_caches.front().originalSize(),
+			_caches.front().frameRate(),
+			ranges::accumulate(
 				_caches,
 				0,
 				std::plus<>(),
diff --git a/lottie/details/lottie_frame_provider_direct.cpp b/lottie/details/lottie_frame_provider_direct.cpp
index dd4a947..a5df4e6 100644
--- a/lottie/details/lottie_frame_provider_direct.cpp
+++ b/lottie/details/lottie_frame_provider_direct.cpp
@@ -95,9 +95,9 @@ bool FrameProviderDirect::load(
 	const auto rate = GetLottieFrameRate(_animation.get(), _quality);
 	const auto count = GetLottieFramesCount(_animation.get(), _quality);
 	return setInformation({
-		.size = QSize(int(width), int(height)),
-		.frameRate = int(rate),
-		.framesCount = int(count),
+		QSize(int(width), int(height)),
+		int(rate),
+		int(count),
 	});
 }
 
diff --git a/lottie/lottie_frame_generator.cpp b/lottie/lottie_frame_generator.cpp
index 5cb6ebd..64d9494 100644
--- a/lottie/lottie_frame_generator.cpp
+++ b/lottie/lottie_frame_generator.cpp
@@ -85,9 +85,9 @@ FrameGenerator::Frame FrameGenerator::renderCurrent(
 		storage.bytesPerLine());
 	_rlottie->renderSync(index * _multiplier, std::move(surface));
 	return {
-		.duration = _frameDuration,
-		.image = std::move(storage),
-		.last = (_frameIndex == _framesCount),
+		_frameDuration,
+		std::move(storage),
+		(_frameIndex == _framesCount),
 	};
 }
 
diff --git a/lottie/lottie_icon.cpp b/lottie/lottie_icon.cpp
index ad02716..cc3cfa6 100644
--- a/lottie/lottie_icon.cpp
+++ b/lottie/lottie_icon.cpp
@@ -141,7 +141,7 @@ private:
 
 Icon::Inner::Inner(int frameIndex, base::weak_ptr<Icon> weak, bool limitFps)
 : _limitFps(limitFps)
-, _current { .index = frameIndex }
+, _current { frameIndex }
 , _weak(weak) {
 }
 
