diff --git a/crl/common/crl_common_object_async.h b/crl/common/crl_common_object_async.h
index 45842ea..29762c8 100644
--- a/crl/common/crl_common_object_async.h
+++ b/crl/common/crl_common_object_async.h
@@ -392,7 +392,7 @@ object_async<Policy, Type>::object_async(Args &&...args)
 	constexpr auto plain_construct = std::is_constructible_v<
 		Type,
 		Args...>;
-	[[maybe_unused]] constexpr auto with_weak_construct = std::is_constructible_v<
+	constexpr auto with_weak_construct = std::is_constructible_v<
 		Type,
 		weak_async<Policy, Type>,
 		Args...>;
