diff --git a/lottie/details/lottie_frame_provider_cached.cpp b/lottie/details/lottie_frame_provider_cached.cpp
index 3a47bc0..3689df5 100644
--- a/lottie/details/lottie_frame_provider_cached.cpp
+++ b/lottie/details/lottie_frame_provider_cached.cpp
@@ -26,9 +26,9 @@ FrameProviderCached::FrameProviderCached(
 		}
 	} else {
 		_direct.setInformation({
-			.size = _cache.originalSize(),
-			.frameRate = _cache.frameRate(),
-			.framesCount = _cache.framesCount(),
+			_cache.originalSize(),
+			_cache.frameRate(),
+			_cache.framesCount(),
 		});
 	}
 }
diff --git a/lottie/details/lottie_frame_provider_cached_multi.cpp b/lottie/details/lottie_frame_provider_cached_multi.cpp
index 0bf5cf1..e075783 100644
--- a/lottie/details/lottie_frame_provider_cached_multi.cpp
+++ b/lottie/details/lottie_frame_provider_cached_multi.cpp
@@ -73,9 +73,9 @@ FrameProviderCachedMulti::FrameProviderCachedMulti(
 	}
 	if (!_direct.loaded()) {
 		_direct.setInformation({
-			.size = _caches.front().originalSize(),
-			.frameRate = _caches.front().frameRate(),
-			.framesCount = ranges::accumulate(
+			_caches.front().originalSize(),
+			_caches.front().frameRate(),
+			ranges::accumulate(
 				_caches,
 				0,
 				std::plus<>(),
diff --git a/lottie/details/lottie_frame_provider_direct.cpp b/lottie/details/lottie_frame_provider_direct.cpp
index 8ca1119..305c7b5 100644
--- a/lottie/details/lottie_frame_provider_direct.cpp
+++ b/lottie/details/lottie_frame_provider_direct.cpp
@@ -93,9 +93,9 @@ bool FrameProviderDirect::load(
 	const auto rate = GetLottieFrameRate(_animation.get(), _quality);
 	const auto count = GetLottieFramesCount(_animation.get(), _quality);
 	return setInformation({
-		.size = QSize(int(width), int(height)),
-		.frameRate = int(rate),
-		.framesCount = int(count),
+		QSize(int(width), int(height)),
+		int(rate),
+		int(count),
 	});
 }
 
diff --git a/lottie/lottie_emoji.cpp b/lottie/lottie_emoji.cpp
index b3bf14c..803575b 100644
--- a/lottie/lottie_emoji.cpp
+++ b/lottie/lottie_emoji.cpp
@@ -70,8 +70,8 @@ EmojiGenerator::Frame EmojiGenerator::renderNext(
 		storage.bytesPerLine());
 	_rlottie->renderSync(index * _multiplier, std::move(surface));
 	return {
-		.image = std::move(storage),
-		.duration = _frameDuration,
+		std::move(storage),
+		_frameDuration,
 	};
 }
 
diff --git a/lottie/lottie_icon.cpp b/lottie/lottie_icon.cpp
index 3d040d5..4373a44 100644
--- a/lottie/lottie_icon.cpp
+++ b/lottie/lottie_icon.cpp
@@ -127,7 +127,7 @@ private:
 };
 
 Icon::Inner::Inner(int frameIndex, base::weak_ptr<Icon> weak)
-: _current{ .index = frameIndex }
+: _current{ frameIndex }
 , _weak(weak) {
 }
 
